File: intermediate-package/writing-integ-tests-for-spring-web-apps/normal-controllers/spring/src/main/java/com/testwithspring/intermediate/user/PersonFinder.java
package com.testwithspring.intermediate.user;

import com.testwithspring.intermediate.common.NotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class PersonFinder {

    private static final Logger LOGGER = LoggerFactory.getLogger(PersonFinder.class);

    private final UserRepository repository;

    @Autowired
    PersonFinder(UserRepository repository) {
        this.repository = repository;
    }

    /**
     * Finds the person information of the user whose id is given as a
     * method parameter.
     * @param userId    The id of the requested user.
     * @return          The found person information.
     * @throws NotFoundException if no person information is found with the provided user id.
     */
    @Transactional(readOnly = true)
    public PersonDTO findPersonInformationByUserId(Long userId) {
        LOGGER.info("Finding person information by using user id: {}", userId);

        PersonDTO found = repository.findPersonInformationById(userId).orElseThrow(
                () -> new NotFoundException(String.format(
                        "No person information found with user id: %d",
                        userId
                ))
        );

        LOGGER.debug("Found person information: {}", found);
        return found;
    }
}

File: master-package/bonus-lessons-2020/multiple-modules-same-data-sets/integration-tests-rest-api/src/main/kotlin/com/testwithspring/master/user/PersonFinder.kt
package com.testwithspring.master.user

import com.testwithspring.master.common.NotFoundException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*


/**
 * Provides a function which queries person information from
 * the database.
 */
@Component
open class PersonFinder(@Autowired private val jdbcTemplate: NamedParameterJdbcTemplate) {

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(PersonFinder::class.java)

        private const val QUERY_FIND_PERSON_INFORMATION = "SELECT u.id as id, " +
                "u.name AS name " +
                "FROM user_accounts u " +
                "WHERE u.id = :id"

        private class PersonInformationRowMapper : RowMapper<PersonDTO> {

            @Throws(SQLException::class)
            override fun mapRow(rs: ResultSet, rowNum: Int): PersonDTO {
                return PersonDTO(
                        name = rs.getString("name"),
                        userId = rs.getLong("id")
                )
            }
        }
    }

    /**
     * Finds the person information information of the specified user.
     * @param   userId  The user id of the user who information will be returned.
     * @return  The "found" person information.
     * @throws  NotFoundException   If no person is found with the given user id.
     */
    @Transactional(readOnly = true)
    open fun findPersonInformationByUserId(userId: Long): PersonDTO {
        LOGGER.info("Finding person information by using the user id: {}", userId)

        try {
            val found: PersonDTO = jdbcTemplate.queryForObject(
                    QUERY_FIND_PERSON_INFORMATION,
                    mapOf(Pair("id", userId)),
                    PersonInformationRowMapper()
            )
            LOGGER.info("Found person information: {}", found)
            return found
        } catch (ex: EmptyResultDataAccessException) {
            LOGGER.error("No person information found with user id: {}", userId)
            throw NotFoundException("No person information found with user id: $userId")
        }
    }
}